core/src/main/java/hudson/util/io/TarArchiver.java

```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
         int mode = IOUtils.mode(file);
         if (mode!=-1)   te.setMode(mode);
         te.setModTime(file.lastModified());
-        if(!file.isDirectory())
-            te.setSize(file.length());
-
-        tar.putArchiveEntry(te);
+        long size = 0;
 
         if (!file.isDirectory()) {
-            FileInputStream in = new FileInputStream(file);
-            try {
-                int len;
-                while((len=in.read(buf))>=0)
-                    tar.write(buf,0,len);
-            } finally {
-                in.close();
+            size = file.length();
+            te.setSize(size);
+        }
+        tar.putArchiveEntry(te);
+        try {
+            if (!file.isDirectory()) {
+                FileInputStream in = new FileInputStream(file);
+                try {
+                    int len;
+                    int read = 0;
+                    while ((len = in.read(buf)) >= 0) {
+                        if (len + read <= size) { // ensure we don't write more bytes than the declared when we created
+                                                  // the entry
+                            tar.write(buf, 0, len);
+                            read += len;
+                        } else {
+                            tar.write(buf, 0, (int) size - read);
+                            break;
+                        }
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
+                    throw e;
+                } finally {
+                    in.close();
+                }
             }
+        } finally { // always close the entry
```

Comment:
Yes, the tar get's closed else were, but in some use cases you will get an `java.io.IOException: This archives contains unclosed entries` when closing the tar in a finally that will mask the original exception. This way we try not to cause another exception. ie: [FilePath:2288](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/FilePath.java#L2288-L2291)